### PR TITLE
xds: use golang protobuf

### PR DESCRIPTION
**Pull request by kyessenov**
Re-creating the PR from upstream branch.
/cc @howardjohn

### CODE REVIEW

**Comment by kyessenov** [comment]
Latest status: Galley can't load standard golang protobufs which are used by mixerclient config and exposed as istio APIs. Turned that path off for now.

**Comment by kyessenov** [comment]
Latest: the workaround for the overlap between galley and xDS protos was to fork the standard protos here and use them exclusively in xDS.

**Comment by kyessenov** [comment]
/retest
The results don't match this PR :/

**Comment by kyessenov** [comment]
Data race looks legitimate:
```
WARNING: DATA RACE
Write at 0x000004c0b8b8 by goroutine 207:
  sync/atomic.StoreInt32()
      /usr/local/go/src/runtime/race_amd64.s:229 +0xb
  github.com/golang/protobuf/proto.(*marshalInfo).size()
      /home/prow/go/pkg/mod/github.com/golang/protobuf@v1.3.2/proto/table_marshal.go:203 +0x326
  github.com/golang/protobuf/proto.makeMessageSliceMarshaler.func1()
      /home/prow/go/pkg/mod/github.com/golang/protobuf@v1.3.2/proto/table_marshal.go:2248 +0xcd
  github.com/golang/protobuf/proto.(*marshalInfo).size()
      /home/prow/go/pkg/mod/github.com/golang/protobuf@v1.3.2/proto/table_marshal.go:183 +0x1b0
  github.com/golang/protobuf/proto.makeMessageSliceMarshaler.func1()
      /home/prow/go/pkg/mod/github.com/golang/protobuf@v1.3.2/proto/table_marshal.go:2248 +0xcd
  github.com/golang/protobuf/proto.(*marshalInfo).size()
      /home/prow/go/pkg/mod/github.com/golang/protobuf@v1.3.2/proto/table_marshal.go:183 +0x1b0
  github.com/golang/protobuf/proto.(*InternalMessageInfo).Size()
      /home/prow/go/pkg/mod/github.com/golang/protobuf@v1.3.2/proto/table_marshal.go:125 +0x73
  github.com/envoyproxy/go-control-plane/envoy/api/v2.(*Listener).XXX_Size()
      /home/prow/go/pkg/mod/github.com/envoyproxy/go-control-plane@v0.9.0/envoy/api/v2/lds.pb.go:96 +0x50
  github.com/golang/protobuf/proto.Marshal()
      /home/prow/go/pkg/mod/github.com/golang/protobuf@v1.3.2/proto/table_marshal.go:2713 +0x3ca
  github.com/golang/protobuf/ptypes.MarshalAny()
      /home/prow/go/pkg/mod/github.com/golang/protobuf@v1.3.2/ptypes/any.go:66 +0x5d
  istio.io/istio/pilot/pkg/proxy/envoy/v2.ldsDiscoveryResponse()
      /home/prow/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/lds.go:79 +0x1c6
  istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).pushLds()
      /home/prow/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/lds.go:35 +0x112
  istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).pushConnection()
      /home/prow/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/ads.go:616 +0x8bc
  istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).StreamAggregatedResources()
      /home/prow/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/ads.go:480 +0x3c23
  github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2._AggregatedDiscoveryService_StreamAggregatedResources_Handler()
      /home/prow/go/pkg/mod/github.com/envoyproxy/go-control-plane@v0.9.0/envoy/service/discovery/v2/ads.pb.go:181 +0xcd
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:1199 +0x1521
  google.golang.org/grpc.(*Server).handleStream()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:1279 +0x12d7
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:710 +0xc8

Previous read at 0x000004c0b8b8 by goroutine 174:
  istio.io/istio/pilot/pkg/networking/core/v1alpha3.(*ConfigGeneratorImpl).buildSidecarOutboundListenerForPortOrUDS()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/listener.go:1310 +0x3440
  istio.io/istio/pilot/pkg/networking/core/v1alpha3.(*ConfigGeneratorImpl).buildSidecarOutboundListeners()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/listener.go:895 +0x14d2
  istio.io/istio/pilot/pkg/networking/core/v1alpha3.(*ConfigGeneratorImpl).buildSidecarListeners()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/listener_builder.go:188 +0x16b
  istio.io/istio/pilot/pkg/networking/core/v1alpha3.(*ConfigGeneratorImpl).BuildListeners()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/v1alpha3/listener.go:276 +0x1bd
  istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).generateRawListeners()
      /home/prow/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/lds.go:50 +0xf3
  istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).pushLds()
      /home/prow/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/lds.go:30 +0x8f
  istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).pushConnection()
      /home/prow/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/ads.go:616 +0x8bc
  istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).StreamAggregatedResources()
      /home/prow/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/ads.go:480 +0x3c23
  github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2._AggregatedDiscoveryService_StreamAggregatedResources_Handler()
      /home/prow/go/pkg/mod/github.com/envoyproxy/go-control-plane@v0.9.0/envoy/service/discovery/v2/ads.pb.go:181 +0xcd
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:1199 +0x1521
  google.golang.org/grpc.(*Server).handleStream()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:1279 +0x12d7
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:710 +0xc8

Goroutine 207 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func1()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:708 +0xb8
  google.golang.org/grpc/internal/transport.(*serverHandlerTransport).HandleStreams()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/internal/transport/handler_server.go:382 +0xb6c
  google.golang.org/grpc.(*Server).serveStreams()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:706 +0x19a
  google.golang.org/grpc.(*Server).ServeHTTP()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:758 +0x166
  istio.io/istio/pilot/pkg/bootstrap.(*Server).initSecureGrpcServer.func2()
      /home/prow/go/src/istio.io/istio/pilot/pkg/bootstrap/server.go:1114 +0xb5
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2007 +0x51
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2802 +0xce
  net/http.initNPNRequest.ServeHTTP()
      /usr/local/go/src/net/http/server.go:3374 +0xfc
  net/http.(*initNPNRequest).ServeHTTP()
      <autogenerated>:1 +0xa6
  net/http.Handler.ServeHTTP-fm()
      /usr/local/go/src/net/http/server.go:86 +0x64
  net/http.(*http2serverConn).runHandler()
      /usr/local/go/src/net/http/h2_bundle.go:5713 +0xac

Goroutine 174 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func1()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:708 +0xb8
  google.golang.org/grpc/internal/transport.(*serverHandlerTransport).HandleStreams()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/internal/transport/handler_server.go:382 +0xb6c
  google.golang.org/grpc.(*Server).serveStreams()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:706 +0x19a
  google.golang.org/grpc.(*Server).ServeHTTP()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:758 +0x166
  istio.io/istio/pilot/pkg/bootstrap.(*Server).initSecureGrpcServer.func2()
      /home/prow/go/src/istio.io/istio/pilot/pkg/bootstrap/server.go:1114 +0xb5
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2007 +0x51
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2802 +0xce
  net/http.initNPNRequest.ServeHTTP()
      /usr/local/go/src/net/http/server.go:3374 +0xfc
  net/http.(*initNPNRequest).ServeHTTP()
      <autogenerated>:1 +0xa6
  net/http.Handler.ServeHTTP-fm()
      /usr/local/go/src/net/http/server.go:86 +0x64
  net/http.(*http2serverConn).runHandler()
      /usr/local/go/src/net/http/h2_bundle.go:5713 +0xac
==================
```

File: istioctl/pkg/auth/listener.go [hunk 1]
```diff
@@ -86,17 +90,30 @@ func getHTTPConnectionManager(filter *listener.Filter) *hcm_filter.HttpConnectio
...
 	case *hcm_filter.HttpFilter_TypedConfig:
-		if err := types.UnmarshalAny(c.TypedConfig, out); err != nil {
+		if err := ptypes.UnmarshalAny(c.TypedConfig, out); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
+func StructToGoGoMessage(pbst *structpb.Struct, out proto.Message) error {
```
Comment (howardjohn) [inline istioctl/pkg/auth/listener.go]: unused function

**Review by howardjohn — APPROVED** [review]

**Review by ayj — APPROVED** [review]

**Review by bianpengyuan — APPROVED** [review]

**Review by incfly — APPROVED** [review]

**Review by esnible — APPROVED** [review]

File: istioctl/pkg/auth/listener.go (hunk 1 above)
Comment (kyessenov) [reply istioctl/pkg/auth/listener.go]: do you know why linter is not catching this? it used to be able to detect unused functions.

Comment (howardjohn) [reply istioctl/pkg/auth/listener.go]: Probably public functions don't get the check

**Review by rshriram — APPROVED** [review]

**Comment by geeknoid** [comment]
@kyessenov Where's the logic located to generate those .pb.go files located? I couldn't find it.

**Comment by kyessenov** [comment]
I didn't add generation script. The logic is to just run `protoc --go_out` in the build container.

**Comment by geeknoid** [comment]
@kyessenov Please add the necessary logic to regenerate those files via go:generate. It's not acceptable to have permanently checked in generated code that can't be auto-updated.

Thanks.

**Comment by kyessenov** [comment]
Ok, but I wasn't the first :) There were already a bunch of committed generated file in security/protos.

**Pull request merged**